course4/lecture4_en.md

```diff
@@ -0,0 +1,306 @@
+
+```moonbit
+Cons('H', Cons('i', Cons('!', Nil)))
+```
+
+The definition of tuples is represented by expressions enclosed in parentheses and separated by commas. The types are also specified using the same syntax, as in the definition of personal identity information here: `("Bob", 2023, 10, 24): (String, Int, Int, Int)`.
+
+Tuple members are accessed by indexes, starting from 0 on the left. For example, `(2023, 10, 24).0 == 2023`.
+
+# Cartesian Product
+
+You may be familiar with the concept of the Cartesian product. The Cartesian product of two sets is a set where all elements are ordered pairs formed by the elements of the original two sets. For example, the Cartesian product of the set of card suits and the numbers 1 to 52 forms all possible ordered pairs of suits and numbers. 
+
+{ ♥️ ♦️ ♠️ ♣️ } $\times \{ n \in \mathbb{N} | 1 \leq n \leq 52 \}$ 
+
+Tuples, on the other hand, go beyond the Cartesian product of two sets; they represent the Cartesian product of multiple sets, making them more accurately termed as ordered sets. Consequently, tuples are also known as product types. You might wonder if there are sum types alongside product types. We will introduce sum types and explore the concepts of **zero** and **one** later.
+
+# Structures
+
+The problem is that it is hard to understand the data represented by tuples. For example, `(String, Int)` – does it represent a person's name and age, or a person's name and phone number, or perhaps address and email?
+
+Structures allow us to assign **names** to the data, both to the entire type and to each field individually. For instance:
+
+- `struct PersonalInfo { name: String; age: Int }`
+- `struct ContactInfo { name: String; telephone: Int }`
+- `struct AddressInfo { address: String; postal: Int }`
+
+We can clearly understand the information about the data and the meaning of each corresponding field by names.
+
+The syntax of defining a structure is `struct <struct_name> { <field_name>: <type> ; ... }`. For example: `struct PersonalInfo { name: String; age: Int }`. As we have mentioned, the semicolons can be omitted when the fields are on separate lines.
+
+Definitions of the values of a structure are enclosed in braces, with each field assigned a value: `<field_name>: <value>`; each field assignment is followed by a comma. For instance: `let info: PersonalInfo = { name: "Moonbit", age: 1, }`. Note the comma after each value. It can be omitted if the last comma is directly followed by the closing brace without a newline. The order of fields does not matter, for example, `{ age: 1, name: "Moonbit" }`. If two structures happen to have exactly the same field names and types, they cannot be distinguished solely by the defined values. In such cases, we can differentiate them by adding a type declaration after the structure's value, such as `: A`.
```

Comment:
Maybe a more complete example than just `: A` would be better?

Reply:
I was thinking about an example where you may have

`let jack = ({ ... } : PersonalInfo)` just to demonstrate how to specify the type of a value (rather than the type of an identifier).